firmware/examples/Post/source/main.cpp

```diff
@@ -0,0 +1,287 @@
+#include <cstdarg>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/gpio.hpp"
+#include "L1_Drivers/i2c.hpp"
+#include "L1_Drivers/ssp.hpp"
+#include "L2_Utilities/log.hpp"
+#include "L2_Utilities/status.hpp"
+#include "L3_HAL/onboard_led.hpp"
+#include "L3_HAL/ssd1306.hpp"
+#include "config.hpp"
+
+constexpr uint8_t kFirstI2cAddress = 0x08;
+constexpr uint8_t kLastI2cAddress  = 0x78;
+
+constexpr uint8_t kAccelerometerAddress = 0x1C;
+constexpr uint8_t kGestureAddress       = 0x39;
+constexpr uint8_t kTemperatureAddress   = 0x31;
+
+I2c i2c;
+
+bool ExternalFlashTest();
+bool I2cTest();
+void OnBoardLedTest(bool gpio_test, bool i2c_test, bool flash_test);
+bool GpioFunctionalityTest();
+bool GestureSensorTest();
+bool AccelerometerTest();
+bool TemperatureSensorTest();
+void OledTest();
+
+int main(void)
+{
+  OledTest();
+  bool gpio_pass  = GpioFunctionalityTest();
+  bool i2c_pass   = I2cTest();
+  bool flash_test = ExternalFlashTest();
+  OnBoardLedTest(gpio_pass, i2c_pass, flash_test);
+
+  // LOG_INFO("Running loop.");
+  Delay(5000);
+  return 0;
+}
+
+void OledTest()
+{
+  Ssd1306 display;
+  LOG_INFO("Starting OLED Hardware Test...");
+
+  LOG_INFO("Initializing OLED Hardware Test...");
+  display.Initialize();
+
+  LOG_INFO("Clearing Screen...");
+  display.Clear();
+
+  LOG_INFO("Clearing Screen finished. Waiting 1 second...");
+
+  Delay(1000);
+  display.Update();
+  LOG_INFO("Inverting screen...");
+  display.InvertScreenColor();
+}
+
+bool ExternalFlashTest()
+{
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting External Flash Test...\n\n");
+  Ssp ssp2(Ssp::Peripheral::kSsp2);
+  Gpio cs(1, 10);
+  cs.SetAsOutput();
+  cs.SetHigh();
+
+  ssp2.SetPeripheralMode(Ssp::MasterSlaveMode::kMaster, Ssp::FrameMode::kSpi,
+                         Ssp::DataSize::kEight);
+  ssp2.SetClock(false, false, 1, 48);
+  ssp2.Initialize();
+  Delay(1000);
+
+  uint8_t array[5];
+  cs.SetLow();
+  Delay(1);
+  // Read Manufacturer ID
+  ssp2.Transfer(0x9F);
+  array[0] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[1] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[2] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[3] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  LOG_INFO("Returned 0x%02X 0x%02X 0x%02X 0x%02X", array[0], array[1], array[2],
```

Comment:
Wait, wait, wait. What are you testing here? What if the bytes coming back area all zeros or all `0xff`? You need to check the values to verify that the test actually worked. Remember that the people in the factory will not be reading the output from the console. And if we made them do that, that would be very confusing to have to look for a specific pattern of bytes, when we could just simply check them in the code.